Add support for non-optimistic updates (+ API versioning)

khamidou opened the pull request:
**Summary:**
This PR adds a new way to update API objects. Instead of updating API objects before pushing the changes to the backend server (aka optimistic updates), we now push the changes to the backend and wait for the sync engine to pick up the changes. This solves a lot of problems we've had previously with folder and labels, where our datastore and the remote server would get out of sync.

For example, in the case of a folder rename, PUTting changes to `/folders/my_id` will return the title of the original folder. The new name will be returned only once we've carried the folder rename operation. This is a breaking API change, which could be very confusing to our long-time API users.

To keep backward compatibility, I've introduced a very basic API versioning scheme, similar to what Github does with its API. API versions are dates. You can set the API version you'd like to use by setting the "X-Api-Version" header in your request. If you don't set this header, we'll default to the version set for your developer dashboard account (changes forthcoming).

This diff is pretty close to review but I need to take a final pass to make sure everything fits together.

**Test Plan:**
Make a final manual test pass using IMAP and Exchange accounts.

**Reviewers:**
spang, bengotow

bengotow commented on inbox/actions/base.py:
```diff
@@ -71,26 +71,30 @@ def create_folder(crispin_client, account_id, category_id):
     remote_delete_folder(crispin_client, account_id, category_id)
 
 
+def create_label(crispin_client, account_id, category_id):
```
This is a bit of a sidenote, but what's the point of this layer of indirection? It doesn't seem like any of the functions in this file really do anything anymore.

khamidou replied:
They let you add behaviour for actions regardless of the provider (e.g: https://github.com/nylas/sync-engine/blob/1d595f2a725b6ee4f2fce8185fc7efa7a07ecac3/inbox/actions/base.py#L121)

bengotow commented on inbox/api/ns_api.py:
```diff
@@ -625,20 +641,24 @@ def folder_label_delete_api(public_id):
-        category.deleted_at = deleted_at
-        for label in category.labels:
-            label.deleted_at = deleted_at
+        if g.api_version == 1:
```
This might be cleaner if we translate the API version into API feature flags, and then check those. The comment (`#Optimistically update labels`) wouldn't be as necessary if we could say `if g.api_features.optimistic_updates:`

khamidou replied:
Good point, will add this.

bengotow commented on inbox/api/ns_api.py:
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
         # db. With MySQL, this means that the column will be 64k.
         # Drop the latest participants until it fits in the column.
-        while len(json.dumps(cancelled_participants)) > 63000:
+        while len(json.dumps(cancelled_participants)) > 2**15:
```
Now 32k, not 63k? Why make the max less than the max size of the column?

khamidou replied:
Hmm, looks like a mistake. Fixing.

bengotow commented on inbox/api/ns_api.py:
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
-    for attr in Event.API_MODIFIABLE_FIELDS:
-        if attr in data:
-            setattr(event, attr, data[attr])
+    if g.api_version == 1:
```
Another approach might be to say `should_update_optimistically(g.api_version):`

khamidou replied:
Fixed!

bengotow commented on inbox/api/ns_api.py:
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
+        g.db_session.commit()
+
+        # Don't sync back updates to autoimported events.
+        # FIXME @karim: document this and forbid it.
```
Rather than having this just silently not do anything can we go ahead and return a 400 with a helpful error message? I'd rather people discover the limitations of our API while using it than discover that it sometimes doesn't work for an unknown reason.

khamidou replied:
Good point.

khamidou replied:
Good point, fixing.

bengotow commented on inbox/api/ns_api.py:
```diff
@@ -863,26 +883,43 @@ def event_update_api(public_id):
+    else:
+        # This isn't an optimistic update, so we need to store the
+        # updated attributes inside the ActionLog entry.
+        kwargs = dict(calendar_uid=event.calendar.uid,
```
There's some code duplication in here that would be nice to avoid. I'm not sure why this input error, for example, is only checked when the update isn't optimistic, and not when it is:

```
if len(json.dumps(kwargs)) > 2**16 - 12:
       raise InputError('Event update too big --- please break it in parts.')
```

Maybe we could generate `kwargs`, run the emailed_events calendar check, and the input length check /before/ we decide what to do?

khamidou replied:
Basically there's two flows here:
1. Optimistic updates. We update the event object in-place and save it. The ActionLog worker will use this event object to update the remote backend.
2. Non-optimistic updates. We save the changes we've received to the ActionLog's extra field (a JSON column). Because this column is a TEXT column, it's limited to 64k chars, so we need to check there's enough space for everything.

bengotow commented on inbox/api/update.py:
```diff
@@ -9,26 +9,28 @@
 
 
-def update_message(message, request_data, db_session):
+def update_message(message, request_data, db_session, optimistic=True):
```
Just a preference, but I'd prefer to default optimistic to `None` and `assert(optimistic is not None)` to make sure that this diff isn't overlooking any places this function is called. Defaulting to true could mean we don't find a bug somewhere where the code hasn't been updated and should be passing False.

bengotow commented on inbox/api/update.py:
```diff
@@ -9,26 +9,28 @@
         if folder is not None:
-            update_message_folder(message, db_session, folder)
+            update_message_folder(message, db_session, folder,
+                                  optimistic=optimistic)
```
(The above would also be nice, because it'd ensure we never forget to do this.) Leaving off `optimistic=optimistic` would cause the wrong behavior here and would be very subtle.

bengotow commented:
LGTM! Just added a few comments inline.

It'd also be really nice if we could change the API response for non-optimistic PUTs and POSTs. I think it'll be confusing to users that they say unread: false, and we return a Thread with unread: true still. When we implemented this in K2, we made the PUT and POST requests return a task object, which you could query to see if the operation had completed. The task objects were also in the delta stream, so you knew when a particular change had been committed to the provider. It could be a separate PR, but that would allow us to finally change the status indicator in N1 (eg "Applying labels...") so that it reflected when the sync engine finished, not the API request.

grinich commented:
Do folks have strong feelings on having an integer version number versus a date? The way I've seen Stripe and Clearbit do their gatekeeper versioning is based on a date param that gets sent. More here:

https://stripe.com/docs/upgrades

http://blog.clearbit.com/versioning-apis/

khamidou commented:
I have no preferences about it. The latest version of this branch uses a date field instead of an integer version.

grinich commented:
Cool, I personally like the date field since it's more explicitly about when the change happened. We should probably also have an API changelog recorded somewhere.

spang commented on inbox/api/srv.py:
```diff
@@ -130,4 +130,5 @@ def logout():
 
 
 app.register_blueprint(ns_api)
+app.register_blueprint(ns_api, url_prefix='/v2')
```
...what is this for? cruft before we decided to go with gating rather than changing the URL?

khamidou replied:
Yup. Removed it.

spang commented on tests/api/test_events.py:
```diff
@@ -152,6 +153,40 @@ def test_api_update_title(db, api_client, calendar, default_account):
+    }
+
+    e_resp = api_client.post_data('/events', e_data,
+                                  headers={ "X-API-Version": API_VERSIONS[1] })
```
Is this the same header that other APIs which do this use?

khamidou replied:
No. Stripe uses `Stripe-Version`. Clearbit uses `Api-Version`. Apparently, it's not recommended anymore to append `X-` to HTTP headers, so I'll just go with `Api-Version`.

spang commented on inbox/transactions/actions.py:
```diff
@@ -222,9 +222,9 @@ def _process_log(self):
-                self.log.info('Syncback namespace_ids count', shard_id=key,
-                              process=self.process_number,
-                              num_namespace_ids=len(namespace_ids))
+                #self.log.info('Syncback namespace_ids count', shard_id=key,
```
Is this related? Not sure what it has to do with this patch

khamidou replied:
Oh I commented those because they spam the console and make it super annoying to debug API requests.

spang commented on inbox/transactions/actions.py:
```diff
@@ -239,10 +239,10 @@ def _process_log(self):
-                      process=self.process_number,
-                      duration=(after - before).total_seconds(),
-                      idle_workers=self.num_idle_workers)
+        #self.log.info('Syncback completed one iteration',
```
or this

spang commented on inbox/api/ns_api.py:
```diff
@@ -300,7 +319,12 @@ def thread_api_update(public_id):
-    update_thread(thread, data, g.db_session)
+
+    if g.api_features.optimistic_updates:
+        update_thread(thread, data, g.db_session, True)
```
any reason we're not just passing `g.api_features.optimistic_updates` in as the argument to `update_thread()`. Seems kind of weird to put a conditional on a boolean and then explicitly pass in the exact same boolean when the same thing could be accomplished in a single line of code

khamidou replied:
Good point.

spang commented:
This in general LGTM. Can you explain the folder and calendar pieces of this? They're pretty complex and I don't quite understand how they work.

khamidou commented:
Sure. 

Concerning folders, the main issue we're having is that we're slow to detect folder renames/deletes (see: https://phab.nylas.com/T6988), so to save time, we simply update the folder once we've got confirmation from the IMAP server that the rename went correctly.

We have a similar problem for event updates --- because of the way ActiveSync works, we _have_ to update the event once the server acks our request. To make this work, we store the updated fields in the `ActionLog` entry's `extra_args` fields (it's a JSON TEXT field, which is why there's some logic to make sure updates aren't bigger than 64k). Note that we're doing the same thing for Google Calendar --- we could have waited for the server to notify us but I preferred to share the same logic for event updates.

I've updated the diff to address your other comments:
- you can now set the API version by using the `Api-Version` HTTP header
- cleaned up the cruft

spang commented:
LGTM! Let's take this to staging! :)
